src/compiler/scala/tools/nsc/typechecker/TypeDiagnostics.scala

```diff
@@ -527,8 +527,10 @@ trait TypeDiagnostics {
           // Only record type references which don't originate within the
           // definition of the class being referenced.
           if (t.tpe ne null) {
-            for (tp <- t.tpe if !treeTypes(tp) && !currentOwner.ownerChain.contains(tp.typeSymbol)) {
-              tp match {
+            for (tp <- t.tpe if !treeTypes(tp)) {
+              val alias = tp.isInstanceOf[AliasTypeRef]
```

Comment:
Is this the best one can do?

Reply:
`tp.typeSymbolDirect.isAliasType`?

Reply:
That pulls in too much, it turns out. I tried various `isNonClassType` etc, so mostly I don't actually understand the API.

Reply:
> That pulls in too much

Can you give an example?

Reply:
In constructors you see:
```
[log typer] ALIAS C.super.type/class scala.reflect.internal.Types$UniqueSuperType/class C/type AnyRef primary constructor C
[log typer] C.super.type referenced from primary constructor C
```
with the false negative for `object X { private class C }`. So the `AnyRef` is the `Direct.isAliasType`, defined as an alias of `Object`.

I guess it's sufficient to also check `isLocal || isPrivate`.